lib/vfs.js

```diff
@@ -1,65 +1,93 @@
+"use strict";
+
 var fs = require("fs");
+var utils = require("./utils");
+
+function ensureDirectory(path) {
+    if ("/" !== path[path.length - 1]) {
+        path += "/";
+    }
+
+    return path;
+}
 
 module.exports = function vfs(request) {
     return {
         getFile: function getFile(path, cb) {
-            request("/api/vfs/" + path, function(err, response, body) {
-                if (err) return cb(err);
+            var url = "/api/vfs/" + path;
+            var action = "getting file with path " + path;
+
+            request(url, function (err, response, body) {
+                err = err || utils.resolveHttpError(response, "Error " + action + ".");
```

Comment:
Ah, makes sense.

What if we define an intermediate `utils.createBodyCallback` and then dependency chain `createJsonCallback -> createBodyCallback -> createCallback`?

Something like:

```javascript
function createCallback(cb, action) {
    return function (err, response) {
        err = err || resolveHttpError(response, "Error " + action + ".");
        if (err) return cb(err);

        cb(null, response);
    };
}

function createBodyCallback(cb, action) {
    return createCallback(function (err, response) {
        if (err) return cb(err);

        cb(null, response.body, response);
    }, action);
}

function createJsonCallback(cb, action) {
    return createBodyCallback(function (err, body, response) {
        if (err) return cb(err);

        try {
            var data = JSON.parse(body);
        } catch (err) {
            return cb(err);
        }

        cb(null, data, response);
    }, action);
}
```